library/red-black/tree.lisp

```diff
@@ -0,0 +1,510 @@
+        ((Tuple (Black) (Black)) True)
+        ((Tuple (DoubleBlack) (DoubleBlack)) True)
+        ((Tuple (NegativeBlack) (NegativeBlack)) True)
+        (_ False))))
+
+  (declare color-plus-black (Color -> Color))
+  (define (color-plus-black c)
+    (match c
+      ((DoubleBlack) (error "cannot add black to double-black"))
+      ((Black) DoubleBlack)
+      ((Red) Black)
+      ((NegativeBlack) Red)))
+
+  (declare color-minus-black (Color -> Color))
+  (define (color-minus-black c)
+    (match c
+      ((DoubleBlack) Black)
+      ((Black) Red)
+      ((Red) NegativeBlack)
+      ((NegativeBlack) (error "cannot subtract black from negative-black"))))
+
+  (define-type (Tree :elt)
+    "A red-black balanced binary tree, sorted by `<=>' and unique by `=='."
+    ;; exported; an empty tree.
+    ;; considered black for the purpose of the invariants.
+    Empty
+
+    ;; unexported; a tree with at least one element, and possibly children.
+    ;; (Brach clr less elt right)
+    ;; every element of LESS is less than ELT, and every element of RIGHT is greater than ELT.
+    (Branch Color (Tree :elt) :elt (Tree :elt))
+
+    ;; unexported: a double-black leaf node. intermediate stage during deletion; will never exist outside of a
+    ;; `remove' operation.
+    DoubleBlackEmpty
+    )
+
+  ;;; color operations
+
+  (declare tree-plus-black ((Tree :elt) -> (Tree :elt)))
+  (define (tree-plus-black tre)
+    (match tre
+      ((Empty) DoubleBlackEmpty)
+      ((Branch c left pivot right) (Branch (color-plus-black c) left pivot right))
+      ((DoubleBlackEmpty) (error "cannot add black to double-black empty tree"))))
+
+  (declare tree-minus-black ((Tree :elt) -> (Tree :elt)))
+  (define (tree-minus-black tre)
+    (match tre
+      ((Empty) (error "cannot subtract black from empty tree"))
+      ((Branch c left pivot right) (Branch (color-minus-black c) left pivot right))
+      ((DoubleBlackEmpty) Empty)))
+
+  (declare as-black ((Tree :elt) -> (Tree :elt)))
+  (define (as-black tre)
+    (match tre
+      ((Empty) Empty)
+      ((DoubleBlackEmpty) Empty)
+      ((Branch _ left elt right) (Branch Black left elt right))))
+
+  (declare tree-double-black? ((Tree :elt) -> Boolean))
+  (define (tree-double-black? tre)
+    (match tre
+      ((Branch clr _ _ _) (== clr DoubleBlack))
+      ((Empty) False)
+      ((DoubleBlackEmpty) True)))
+
+  ;;; searching trees
+
+  (declare lookup ((Ord :elt) => ((Tree :elt) -> :elt -> (Optional :elt))))
+  (define (lookup haystack needle)
+    "If HAYSTACK contains an element `==' to NEEDLE, return it."
+    (match haystack
+      ((Empty) None)
+      ((Branch _ left elt right)
+       (match (<=> needle elt)
+         ((LT) (lookup left needle))
+         ((EQ) (Some elt))
+         ((GT) (lookup right needle))))))
+
+  ;;; inserting into and replacing elements of trees
+
+  (declare balance (Color -> (Tree :elt) -> :elt -> (Tree :elt) -> (Tree :elt)))
+  (define (balance clr left elt right)
```

Comment:
@_@

Reply:
`balance` is called by the insertion algorithm.